src/CoreWCF.Primitives/tests/Extensibility/InstanceProviderTests.cs

```diff
@@ -95,7 +95,11 @@ public void ReleaseInstance(InstanceContext instanceContext, object instance)
 
         public async Task WaitForReleaseAsync(TimeSpan timeout)
         {
-            (await _asyncLock.TakeLockAsync(timeout))?.Dispose();
+            var relaser = (await _asyncLock.TakeLockAsync(timeout))?.DisposeAsync();
+            if (relaser != null)
```

Comment:
Ah, it makes sense why you added the timeout now. The failure mode of this test should be an exception from the WaitForReleaseAsync method. We expect the release to happen, and the timeout is to enable giving up on that. This test could easily use a semaphoreslim instead similar to the non-reentrant lock you introduced earlier.